vJoySerialFeeder/SerialProtocols/CrsfReader.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System;
+using System.Collections;
+using System.IO.Ports;
+using System.Windows.Forms;
+
+namespace vJoySerialFeeder
+{
+	public class Crc8
+	{
+		private byte[] _table = new byte[256];
+
+		public Crc8(byte poly)
+		{
+			GenerateTable(poly);
+			Out = 0;
+		}
+
+		public void Add(byte b)
+		{
+			_out = _table[_out ^ b];
+		}
+
+		private byte _out = 0;
+		public byte Out
+		{
+			get { return _out;  }
+			set { _out = 0;  }
+		}
+
+		private void GenerateTable(byte poly)
+		{
+			for (uint i = 0; i < 256; ++i)
+			{
+				uint curr = i;
+				for (uint j = 0; j < 8; ++j)
+				{
+					if ((curr & 0x80) != 0)
+					{
+						curr = (curr << 1) ^ poly;
+					}
+					else
+					{
+						curr <<= 1;
+					}
+				}
+
+				_table[i] = (byte)curr;
+			}
+		}
+	}
+
+	public class CrsfReader : SerialReader
+	{
+		const int CRSF_PAYLOAD_SIZE_MAX = 62;
+		const int CRSF_SUBSET_RC_CHANNELS_PACKED_RESOLUTION = 11; // 11 bits per channel
+		const uint CRSF_SUBSET_RC_CHANNELS_PACKED_MASK = 0b11111111111; // 11 bits, get it?!
+
+		// AddressTypes
+		const byte CRSF_ADDRESS_FLIGHT_CONTROLLER = 0xC8; // 200
+		// FrameTypes
+		const byte CRSF_FRAMETYPE_LINK_STATISTICS = 0x14;
+		const byte CRSF_FRAMETYPE_RC_CHANNELS_PACKED = 0x16;
+
+		private Crc8 crcCalc = new Crc8(0xD5);
+
+		public override void Start()
+		{
+			serialPort.ReadTimeout = 500;
+			Buffer.FrameLength = CRSF_PAYLOAD_SIZE_MAX + 2;
+		}
+
+		public override void Stop()
+		{
+		}
+
+		public override int ReadChannels()
+		{
+			int retVal = 0;
+			byte len = Buffer[1];
+			byte addr = Buffer[0];
+			//System.Diagnostics.Debug.WriteLine("CRSF[{0}] len={1}", addr, len);
+			if (len < 1 || len > CRSF_PAYLOAD_SIZE_MAX)
+			{
+				Buffer.Slide(2);
```

Comment:
Here it's preferable to be `Buffer.Slide(1)`. As a whole this should happen fairly rarely. Maybe if we connect the serial port in the middle of the frame transmission. Then the point is to increment byte by byte until we catch a valid frame. If we slide two bytes at a time theoretically we can miss the beginning of the next valid frame (if it's say, the next byte).